core/src/main/java/org/elasticsearch/repositories/blobstore/ChecksumBlobStoreFormat.java

```diff
@@ -0,0 +1,217 @@
+    }
+
+    /**
+     * @param codec          codec name
+     * @param blobNameFormat format of the blobname in {@link String#format} format
+     * @param reader         prototype object that can deserialize T from XContent
+     * @param compress       true if the content should be compressed
+     */
+    public ChecksumBlobStoreFormat(String codec, String blobNameFormat, FromXContentBuilder<T> reader, ParseFieldMatcher parseFieldMatcher, boolean compress) {
+        this(codec, blobNameFormat, reader, parseFieldMatcher, compress, DEFAULT_X_CONTENT_TYPE);
+    }
+
+    /**
+     * Reads blob with specified name without resolving the blobName using using {@link #blobName} method.
+     *
+     * @param blobContainer blob container
+     * @param blobName blob name
+     * @return
+     * @throws IOException
+     */
+    public T readBlob(BlobContainer blobContainer, String blobName) throws IOException {
+        try (InputStream inputStream = blobContainer.openInput(blobName)) {
+            byte[] bytes = ByteStreams.toByteArray(inputStream);
+            final String resourceDesc = "ChecksumBlobStoreFormat.readBlob(blob=\"" + blobName + "\")";
+            try (ByteArrayIndexInput indexInput = new ByteArrayIndexInput(resourceDesc, bytes)) {
+                CodecUtil.checksumEntireFile(indexInput);
+                CodecUtil.checkHeader(indexInput, codec, VERSION, VERSION);
+                long filePointer = indexInput.getFilePointer();
+                long contentSize = indexInput.length() - CodecUtil.footerLength() - filePointer;
+                BytesReference bytesReference = new BytesArray(bytes, (int) filePointer, (int) contentSize);
+                return read(bytesReference);
+            } catch (CorruptIndexException | IndexFormatTooOldException | IndexFormatTooNewException ex) {
+                // we trick this into a dedicated exception with the original stacktrace
+                throw new CorruptStateException(ex);
+            }
+        }
+    }
+
+    /**
+     * Writes blob in atomic manner with resolving the blob name using {@link #blobName} and {@link #tempBlobName} methods.
+     * <p/>
+     * The blob will be compressed and checksum will be written if required.
+     *
+     * Atomic move might be very inefficient on some repositories. It also cannot override existing files.
+     *
+     * @param obj           object to be serialized
+     * @param blobContainer blob container
+     * @param name          blob name
+     * @throws IOException
+     */
+    public void writeAtomic(T obj, BlobContainer blobContainer, String name) throws IOException {
+        String blobName = blobName(name);
+        String tempBlobName = tempBlobName(name);
+        writeBlob(obj, blobContainer, tempBlobName);
+        try {
+            blobContainer.move(tempBlobName, blobName);
+        } catch (IOException ex) {
+            // Move failed - try cleaning up
+            blobContainer.deleteBlob(tempBlobName);
```

Comment:
Maybe we should inject a logger so we can log something here? Otherwise we silently fail the writeAtomic?